File: backend/app/schemas/assistant_suite.py
"""ARCH-39 — request and response shapes for conversation sessions."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class ConversationSessionSummary(BaseModel):
    id: uuid.UUID
    title: str
    kind: Literal["workspace", "document"]
    scope_mode: Literal["WORKSPACE", "SELECTED", "DOCUMENT"]
    work_item_id: Optional[uuid.UUID] = None
    document_title: Optional[str] = None
    scope_document_count: int = 0
    model_override: Optional[str] = None
    pinned: bool = False
    archived: bool = False
    message_count: int = 0
    created_at: datetime
    last_message_at: Optional[datetime] = None


class ConversationSessionUpdate(BaseModel):
    """Every field optional; only the ones present change."""

    title: Optional[str] = Field(default=None, min_length=1, max_length=255)
    pinned: Optional[bool] = None
    archived: Optional[bool] = None
    model_override: Optional[str] = Field(default=None, max_length=96)
    clear_model_override: bool = False


class ConversationScopeRead(BaseModel):
    mode: Literal["WORKSPACE", "SELECTED", "DOCUMENT"]
    work_item_ids: list[uuid.UUID] = Field(default_factory=list)
    max_documents: int


class ConversationScopeUpdate(BaseModel):
    mode: Literal["WORKSPACE", "SELECTED"]
    work_item_ids: list[uuid.UUID] = Field(default_factory=list, max_length=100)


class AssistantModelOption(BaseModel):
    provider: str
    model: str
    is_workspace_default: bool
    input_micros_per_million: Optional[int] = None
    output_micros_per_million: Optional[int] = None
    currency: Optional[str] = None


class PromptTemplateRead(BaseModel):
    id: uuid.UUID
    name: str
    body: str
    created_by_user_id: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PromptTemplateWrite(BaseModel):
    name: str = Field(min_length=1, max_length=80)
    body: str = Field(min_length=1, max_length=8000)


__all__ = [
    "AssistantModelOption",
    "ConversationScopeRead",
    "ConversationScopeUpdate",
    "ConversationSessionSummary",
    "ConversationSessionUpdate",
    "PromptTemplateRead",
    "PromptTemplateWrite",
]

File: frontend/src/types/assistantSuite.ts
/**
 * ARCH-39 — conversation sessions, scope, models and prompt templates.
 * Mirrors app/schemas/assistant_suite.py.
 */

export type ConversationKind = "workspace" | "document";
export type ConversationKindFilter = "all" | ConversationKind;
export type ConversationScopeMode = "WORKSPACE" | "SELECTED" | "DOCUMENT";

export interface ConversationSession {
  readonly id: string;
  readonly title: string;
  readonly kind: ConversationKind;
  readonly scope_mode: ConversationScopeMode;
  readonly work_item_id: string | null;
  readonly document_title: string | null;
  readonly scope_document_count: number;
  readonly model_override: string | null;
  readonly pinned: boolean;
  readonly archived: boolean;
  readonly message_count: number;
  readonly created_at: string;
  readonly last_message_at: string | null;
}

export interface ConversationSessionFilters {
  readonly kind: ConversationKindFilter;
  readonly archived: boolean;
  readonly q: string;
}

export interface ConversationSessionUpdate {
  readonly title?: string;
  readonly pinned?: boolean;
  readonly archived?: boolean;
  readonly model_override?: string;
  readonly clear_model_override?: boolean;
}

export interface ConversationScope {
  readonly mode: ConversationScopeMode;
  readonly work_item_ids: readonly string[];
  readonly max_documents: number;
}

export interface ConversationScopeUpdate {
  readonly mode: "WORKSPACE" | "SELECTED";
  readonly work_item_ids: readonly string[];
}

export interface AssistantModelOption {
  readonly provider: string;
  readonly model: string;
  readonly is_workspace_default: boolean;
  readonly input_micros_per_million: number | null;
  readonly output_micros_per_million: number | null;
  readonly currency: string | null;
}

export interface PromptTemplate {
  readonly id: string;
  readonly name: string;
  readonly body: string;
  readonly created_by_user_id: string | null;
  readonly created_at: string;
  readonly updated_at: string;
}

export interface PromptTemplateWrite {
  readonly name: string;
  readonly body: string;
}

export type ExportFormat = "json" | "markdown";

/** "$0.15 / $0.60 per 1M tokens", or null when the model is unpriced. */
export const formatModelPrice = (option: AssistantModelOption): string | null => {
  if (option.input_micros_per_million === null || option.output_micros_per_million === null) {
    return null;
  }
  const currency = option.currency ?? "USD";
  const format = (micros: number): string =>
    new Intl.NumberFormat(undefined, {
      style: "currency",
      currency,
      maximumFractionDigits: 4,
    }).format(micros / 1_000_000);
  return `${format(option.input_micros_per_million)} in / ${format(
    option.output_micros_per_million,
  )} out per 1M tokens`;
};
